sentry/src/main/java/io/sentry/clientreport/ClientReportRecorderImpl.java

```diff
@@ -0,0 +1,176 @@
+import org.jetbrains.annotations.Nullable;
+
+public final class ClientReportRecorderImpl implements ClientReportRecorder {
+
+  private static volatile @Nullable ClientReportRecorderImpl SHARED_INSTANCE = null;
+
+  public static @NotNull ClientReportRecorderImpl getInstance() {
+    if (SHARED_INSTANCE == null) {
+      synchronized (ClientReportRecorderImpl.class) {
+        if (SHARED_INSTANCE == null) {
+          SHARED_INSTANCE = new ClientReportRecorderImpl();
+        }
+      }
+    }
+
+    return SHARED_INSTANCE;
+  }
+
+  private final ClientReportStorage storage;
+
+  private ClientReportRecorderImpl() {
+    this.storage = new AtomicClientReportStorage();
+  }
+
+  @Override
+  public @NotNull SentryEnvelope attachReportToEnvelope(
+      @NotNull SentryEnvelope envelope, @NotNull SentryOptions options) {
+    @Nullable ClientReport clientReport = resetCountsAndGenerateClientReport();
+    if (clientReport == null) {
+      return envelope;
+    }
+
+    try {
+      options.getLogger().log(SentryLevel.DEBUG, "Attaching client report to envelope.");
+
+      List<SentryEnvelopeItem> items = new ArrayList<>();
+
+      for (SentryEnvelopeItem item : envelope.getItems()) {
+        items.add(item);
+      }
+
+      items.add(SentryEnvelopeItem.fromClientReport(options.getSerializer(), clientReport));
+
+      return new SentryEnvelope(envelope.getHeader(), items);
+    } catch (Throwable e) {
+      options.getLogger().log(SentryLevel.ERROR, e, "Unable to attach client report to envelope.");
+      return envelope;
+    }
+  }
+
+  @Override
+  public void recordLostEnvelope(
+      @NotNull DiscardReason reason,
+      @Nullable SentryEnvelope envelope,
+      @NotNull SentryOptions options) {
+    if (envelope == null) {
+      return;
+    }
+
+    try {
+      for (SentryEnvelopeItem item : envelope.getItems()) {
+        recordLostEnvelopeItem(reason, item, options);
+      }
+    } catch (Throwable e) {
+      options.getLogger().log(SentryLevel.ERROR, e, "Unable to record lost envelope.");
+    }
+  }
+
+  @Override
+  public void recordLostEnvelopeItem(
+      @NotNull DiscardReason reason,
+      @Nullable SentryEnvelopeItem envelopeItem,
+      @NotNull SentryOptions options) {
+    if (envelopeItem == null) {
+      return;
+    }
+
+    try {
+      @NotNull SentryItemType itemType = envelopeItem.getHeader().getType();
+      if (SentryItemType.ClientReport.equals(itemType)) {
+        try {
+          ClientReport clientReport = envelopeItem.getClientReport(options.getSerializer());
+          restoreCountsFromClientReport(clientReport);
+        } catch (Exception e) {
+          options
+              .getLogger()
+              .log(SentryLevel.ERROR, "Unable to restore counts from previous client report.");
+        }
+      } else {
+        recordLostEventInternal(
+            reason.getReason(), categoryFromItemType(itemType).getCategory(), 1L);
+      }
+    } catch (Throwable e) {
```

Comment:
Windering if this try/catch should be within `recordLostEventInternal` only.

Reply:
For now it probably wouldn't make a difference but I sort of wanted this in place so future changes wouldn't throw out.